core/src/main/java/org/apache/accumulo/core/util/Retry.java

```diff
@@ -381,6 +388,48 @@ public NeedsRetryDelay maxRetries(long max) {
       return this;
     }
 
+    @Override
+    public NeedsRetryDelay maxRetriesWithinDuration(Duration duration) {
+      checkState();
+      Preconditions.checkArgument(!duration.isNegative(),
+          "Duration for retries must not be negative");
+      this.retriesForDuration = duration;
+      return this;
+    }
+
+    /**
+     * Calculate the maximum number of retries that can occur within {@link #retriesForDuration}
+     */
+    private void calculateRetriesWithinDuration() {
+      long numberOfRetries = 0;
+      long cumulativeWaitTimeMillis = 0;
+      long currentWaitTimeMillis = initialWait.toMillis();
+      long retriesForDurationMillis = retriesForDuration.toMillis();
+
+      while (cumulativeWaitTimeMillis + currentWaitTimeMillis <= retriesForDurationMillis) {
+
+        cumulativeWaitTimeMillis += currentWaitTimeMillis;
+        numberOfRetries++;
+
+        if (backOffFactor > 1.0) {
+          currentWaitTimeMillis = (long) Math.ceil(currentWaitTimeMillis * backOffFactor);
+        } else {
+          currentWaitTimeMillis += waitIncrement.toMillis();
+        }
+
+        if (currentWaitTimeMillis > maxWait.toMillis()) {
+          currentWaitTimeMillis = maxWait.toMillis(); // Ensure wait time does not exceed maxWait
+        }
+
+        // prevent an infinite loop
+        if (numberOfRetries >= Integer.MAX_VALUE) {
+          break;
+        }
```

Comment:
Is this trying to ensure that `maxRetries` is never set to a negative number? Or is there a different infinite loop possibility?

Reply:
This just makes sure that if an infinite loop occurs for some reason (`numberOfRetries` is incremented at every iteration) the loop will break after a very large number. I am not sure this is the best approach and more or less picked the value to check against at random.

Reply:
Maybe switch it to `long.MAX_VALUE` since `maxRetries` is also a Long. 
The goal is to ensure that `this.maxRetries` is never a negative value since that causes the Retry to be infinite.

Reply:
```suggestion
        // ensure number is less than MAX_VALUE to prevent infinite retries
        if (numberOfRetries >= Long.MAX_VALUE) {
          break;
        }
```

Reply:
Actually I think Long.MAX_VALUE is probably too large here. If we are getting to that number of retries, I think something is wrong and even this simple while loop might take a while to reach that break point.

Reply:
> The break here prevents the while loop in `calculateRetriesWithinDuration()` from being infinite not the retry process itself. I think its a good idea to change it to Long.MAX_VALUE but unless I am misinterpreting something I don't think the comment fits here. The comment could probably be refined to specify that it is referring too the while loop it is in as opposed to the retry process.

Yeah I wasn't thinking clearly. The break doesn't nullify the Retry object creation it just breaks the increment loop. The code would still set `this.maxRetries` to the current number of retries. 

> Actually I think Long.MAX_VALUE is probably too large here. If we are getting to that number of retries, I think something is wrong and even this simple while loop might take a while to reach that break point.

Yep I agree. Plus if we used Long.MaxValue, then it would still be an issue because that's the break condition, not an thrown exception. 

Integer.MAX_VALUE is 2,147,483,647. That's still pretty high for # of retry attempts. 

What if we used something relative to time as the loop limit?
Like `Duration.ofHours(1).toMillis()`?  this would limit number of retries to 3,600,000 or lower.

Reply:
> What if we used something relative to time as the loop limit?
> Like `Duration.ofHours(1).toMillis()`? this would limit number of retries to 3,600,000 or lower.

This makes a lot of sense. Good idea. Added in 2394559